packages/falso/src/lib/core/core.ts

```diff
@@ -11,21 +12,130 @@ type Return<T, O extends FakeOptions> = [O] extends [never]
   : T;
 
 export function fake<T, Options extends FakeOptions>(
-  data: T[] | ((i: number) => T),
-  options?: Options
+  data: T[] | (() => T),
+  options?: Options,
+  comparisonFunction: (item: T, items: T[]) => boolean = primitiveValueIsUnique,
+  comparisonKeys?: string[]
 ): Return<T, Options> {
-  const dataSource = Array.isArray(data) ? () => randElement(data) : data;
+  if (options?.length === 0) {
+    return [] as any;
+  }
 
-  if (options?.length === undefined) {
-    return dataSource(0) as any;
+  if (Array.isArray(data)) {
+    return fakeFromArray(data, options) as any;
   }
 
-  return Array.from({ length: options.length }, (_, index) =>
-    dataSource(index)
+  return fakeFromFunction(
```

Comment:
If these options are only relevant to a function, we can add protection in the fake function using typescript.